### PR TITLE
ROX-20859: Add Konflux build pipeline for Scanner v4 

### PR DESCRIPTION
**Pull request by kylape**
## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scanner/image/scanner/download-mappings.sh
Comment (tommartensen) [inline scanner/image/scanner/download-mappings.sh:8]: Please ensure compatibility with the other usage of this script in https://github.com/stackrox/stackrox/blob/master/scanner/image/scanner/Dockerfile#L11 .

File: .tekton/scanner-v4-pull-request.yaml
Comment (msugakov) [inline .tekton/scanner-v4-pull-request.yaml:229]:
```suggestion
            script: "$(workspaces.source.path)/source/scanner/image/scanner/download-mappings.sh" "$(workspaces.source.path)/source"
```

File: scripts/konflux/fail-build-if-git-is-dirty.sh
Comment (msugakov) [inline scripts/konflux/fail-build-if-git-is-dirty.sh:29]: That's right, our `make tag` does not append `-dirty` if there are only untracked files. I wonder thought if we should allow them here or add to `.gitignore`. I haven't formed an opinion on this.

File: scanner/image/scanner/download-mappings.sh
Comment (msugakov) [inline scanner/image/scanner/download-mappings.sh:24]: [minor] I'd add a message in the end so that it's clear from logs that the script completed as expected.

```suggestion
done

echo "Done"
```

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:34]:
```suggestion
    com.redhat.component="rhacs-scanner-v4-container" \
    com.redhat.license_terms="https://www.redhat.com/agreements" \
    description="This image supports image scanning for RHACS" \
    io.k8s.description="This image supports image scanning for RHACS" \
    io.k8s.display-name="scanner-v4" \
    io.openshift.tags="rhacs,scanner,stackrox" \
    maintainer="Red Hat, Inc." \
    name="rhacs-scanner-v4-rhel8" \
    source-location="https://github.com/stackrox/stackrox" \
```

I also wonder if we should mention `v4` in `summary` and `*description`.

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:54]: This chunk does not exist in the downstream version of the file <https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-scanner-v4/Dockerfile.in?ref_type=heads>. How do you think, did colleagues forget to add it there?

Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile:54]: I should probably double check the code now, given that there has been lots of changes in the last couple of weeks.  And then I'll try to understand why it's not downstream.  It's possible it was forgotten, but I'm not sure.

File: scanner/image/scanner/download-mappings.sh
Comment (kylape) [reply scanner/image/scanner/download-mappings.sh:8]: Done :+1:

File: scanner/image/scanner/konflux.Dockerfile
Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile:54]: Addressed in downstream PR.

File: scanner/image/scanner/konflux.Dockerfile
Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile]: I added something about v4 to summary and description

File: .tekton/scanner-v4-pull-request.yaml
Comment (kylape) [reply .tekton/scanner-v4-pull-request.yaml]: Since this is a single line script in yaml, I made your change and then wrapped the value in single quotations.

File: scripts/konflux/fail-build-if-git-is-dirty.sh
Comment (tommartensen) [reply scripts/konflux/fail-build-if-git-is-dirty.sh:29]: what untracked files did you find? (I assume it's the downloaded ones?)

Comment (tommartensen) [reply scripts/konflux/fail-build-if-git-is-dirty.sh:29]: > it would be good to put it in collector

Collector doesn't embed the version and doesn't have the `fail-build-if-git-is-dirty.sh` script, therefore noop.

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:28]: I'd suggest to move `v4` to reduce confusion.

```suggestion
    description="This image supports image scanning v4 for RHACS" \
```

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:35]: Here too.

```suggestion
    summary="The image scanner v4 for RHACS" \
```

File: .tekton/scanner-v4-pull-request.yaml
Comment (msugakov) [reply .tekton/scanner-v4-pull-request.yaml]: I see

File: scripts/konflux/fail-build-if-git-is-dirty.sh
Comment (msugakov) [reply scripts/konflux/fail-build-if-git-is-dirty.sh:29]: Unresolved because I have an alternative suggestion: instead of excluding untracked files, I suggest creating entries in `.gitignore` for these blobs. This will be more targeted than just ignoring any untracked files.

Not that I see a particular case where extra untracked files will cause harm but it feels cleaner. I'm also not sure how annoying it would be to keep failing on untracked files, it needs some time.

**Review by msugakov — COMMENTED** [review]
Looks good. Let's decide on `fail-build-if-git-is-dirty.sh` vs `.gitignore` [approach](https://github.com/stackrox/stackrox/pull/9864#discussion_r1522002626). Plus it looks like there's a bunch of merge conflicts to resolve (probably task updates concurrently landed on `master`).

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:13]: https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/merge_requests/2745 changed the variable from `RELEASE_TAG` to `BUILD_TAG` which has long history but is a simple change eventually.

The problem is that `CI_VERSION` is not available in Konflux context. We need to use something meaningful here and this warrants a discussion and a decision with the Scanner team.

I'd suggest you set BUILD_TAG to 1.0.0 and create a JIRA with a `# TODO(ROX-....): ...` comment here to follow up.

Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile:13]: If we hard coded BUILD_TAG to 1.0.0, wouldn't we always override the version string?  I thought we'd want to use the output of `make tag` for dev/CI builds.

Comment (msugakov) [reply scanner/image/scanner/konflux.Dockerfile:13]: Yes, we need to solve `make tag`. I thought this could be done in a separate ticket, in a separate PR. But, if you want to make it here - absolutely.
Yet, `$CI_VERSION` is not set in Konflux and I suspect will be empty.

Scanner V4 adds its own vulns update semantics on top of `make tag` thing which we touched in the MR I linked.

Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile:13]: My thinking is that we want to handle the upstream case for now and leave the downstream for later (i.e. create a Jira and a TODO comment).  The upstream case would be to use `make tag` as-is since that gives a more descriptive version string when wanting to know where a particular image came from, for example.

Looking at the [code](https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L198), the version that gets baked into the binary is not actually used to download vulnerabilities from Central, so there should be no issue using `make tag` in regards to downloading vulnerability updates (for now at least).

Perhaps the Jira should include addressing the hard-coded replacement for ROX_VERSION in the vulnerability URL?

Comment (msugakov) [reply scanner/image/scanner/konflux.Dockerfile:13]: > Perhaps the Jira should include addressing the hard-coded replacement for ROX_VERSION in the vulnerability URL?

Mutating side-effects in a function called `validate()` worry me also.

----

> when I deployed a Konflux-built image of scanner, dev was still used as the version string

Maybe that's because the overriding environment variable is still `RELEASE_TAG` and not `BUILD_TAG`?

That's how I trace the workings but maybe I'm wrong because I did not run the code
1. https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L199-L201 will execute if we have `ReleaseBuild`
2. We should have `ReleaseBuild` because of `ENV GOTAGS="release"` here https://github.com/stackrox/stackrox/pull/9864/files#diff-3fe9bb693b9137ea1603abd2ed863434d2e11f55b3ca1fff59d4f06deaeb4b74R12
3. The magic `Version` https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/internal/version/version.go#L1-L4 is set in the Makefile https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/Makefile#L31-L34 which will set it to whatever comes in `$(TAG)`. That, in turn, delegates to the Makefile in the repo root. https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/Makefile#L11-L13 The relevant part in the root Makefile is https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/Makefile#L29-L31 Notice that it does `git describe ...` only when `TAG` is not set. Which will be set in the imported `env.mk` with the value that comes in `BUILD_TAG` https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/make/env.mk#L52-L54
4. In the current `konflux.Dockerfile` `BUILD_TAG` is not set and so `Version` should be a result of `git describe`, that `4.4.x-178-gbc0362419f` you showed. However, what escapes me is why `ENV GOTAGS="release"` does not have effect (in step (2)).

I don't know if there's any condition when this block gets executed https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L202-L204

What happens further in Central with `/api/extensions/scannerdefinitions?version=<ROX_VERSION>` is also interesting although I could not quite parse how the version is applied in https://github.com/stackrox/stackrox/blob/master/central/scannerdefinitions/handler/handler.go

Overall, I feel product version and vulnerability version are unnecessarily entangled which can hardly be beneficial.

I guess, I'd leave it to you to decide about the next step since you're closer to this code, can test and see effects.
The only certain thing form at this point is that `ENV RELEASE_TAG="$CI_VERSION"` has no effect on anything and should be removed.

Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile:13]: I found that the vuln update bundle should be copied to a `1.0.0` version: 

https://github.com/stackrox/stackrox/blob/7a197936f8d9add99e5aef85ded1ba822ff6706c/.github/workflows/scanner-dev-vuln-update.yaml#L46

But when I tried to download it, it wasn't there.  If the vuln bundle is intended to be pushed to that version for a while, then I'd be willing to hard code 1.0.0.

Comment (msugakov) [reply scanner/image/scanner/konflux.Dockerfile:13]: There is some bug: 1.0.0 appears to be an object, not a directory.

```
$ gsutil ls gs://definitions.stackrox.io/v4/vulnerability-bundles/
gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
...
gs://definitions.stackrox.io/v4/vulnerability-bundles/4.4.0/
gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/
```

Notice no `/` at the end.

```
$ gsutil ls -lh gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/  
164.02 MiB  2024-03-20T13:18:37Z  gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/vulns.json.zst
TOTAL: 1 objects, 171991321 bytes (164.02 MiB)
```

```
$ gsutil ls -lh gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
164.02 MiB  2024-03-20T13:18:40Z  gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
TOTAL: 1 objects, 171991321 bytes (164.02 MiB)
```

Maybe those lines should end with trailing slash like
```
        gsutil cp "vulns/vulns.json.zst" "gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/"
        gsutil cp "vulns/vulns.json.zst" "gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0/"
```

Comment (msugakov) [reply scanner/image/scanner/konflux.Dockerfile:13]: I think it should be good enough to hardcode `dev` and add a `# TODO` for now, if it actually works.

Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile]: Alright, I've made that change, but I haven't been able to figure out why the `release` gotag is not being used during compilation.  I created [ROX-23340](https://issues.redhat.com/browse/ROX-23340) as a follow up if you're okay to resolve this and merge.

File: scanner/image/scanner/konflux.Dockerfile
Comment (msugakov) [inline scanner/image/scanner/konflux.Dockerfile:13]:
```suggestion
# TODO(ROX-23335): Properly set the build tag
```

File: .gitleaks.toml
Comment (msugakov) [inline .gitleaks.toml:7]: Is this intended change?

File: .gitignore
Comment (msugakov) [inline .gitignore:84]: Sorry, I just realized now that these will sit in the repo root. As a suggestion, let's put them in some subdirectory to create a context of "scanner build time". I don't have a good directory to suggest but I may have a similar case to solve for `main`.

**Review by msugakov — APPROVED** [review]

File: .gitleaks.toml
Comment (kylape) [reply .gitleaks.toml:7]: When I ran `git merge master` into this branch, my RH pre-commit hook flagged a bunch of these [example value files](https://github.com/stackrox/stackrox/blob/49ab6258de11d4f5e960b1d2bc44a8084f528d6b/image/templates/helm/stackrox-central/values-private.yaml.example#L38) in the helm chart.  This change allowed my merge to be committed.

I suspect why this hasn't been hit before is because I had to resolve a merge conflict after running `git merge master`. After the initial merge command, I addressed the conflict, added the files, and then ran `git merge --continue`, which is when I got the pre-commit errors.

Comment (kylape) [reply .gitleaks.toml:7]: I can separate this out into a different PR if you'd like.

Comment (msugakov) [reply .gitleaks.toml:7]: No, it's fine to keep. I'd suggest to add a comment around that line to show intention, and maybe also scope path (untested):
```suggestion
    # Example value files in Helm charts.
    '''stackrox/image/templates/helm/.*values-private.*\.yaml\.example$''',
```

File: scanner/image/scanner/konflux.Dockerfile
Comment (kylape) [reply scanner/image/scanner/konflux.Dockerfile]: [e64ef27](https://github.com/stackrox/stackrox/pull/9864/commits/e64ef27acf01024b0a5c6d4b159c7d229547170b)

File: .gitleaks.toml
Comment (kylape) [reply .gitleaks.toml]: After failing to reproduce an error by the pre-commit hook, I've just removed this line altogether.  Perhaps there is some kind of bug with the pre-commit hook?  Or maybe there is a problem where it should be flagging those files but it's normally not?  Regardless, removing the line does not cause any issues for me at this point.

File: .gitignore
Comment (kylape) [reply .gitignore:84]: da071cb

**Pull request merged**
